src/Discord/Internal/Types/ScheduledEvents.hs

```diff
@@ -0,0 +1,188 @@
+      , scheduledEventExternalStatus :: ScheduledEventStatus
+      , scheduledEventExternalEntityId :: Maybe ScheduledEventEntityId
+      , scheduledEventExternalCreator :: Maybe User
+      , scheduledEventExternalUserCount :: Maybe Integer
+        -- | FIXME: that's a compatible type, but idealy that's something we want a dedicated type for
+      , scheduledEventExternalImage :: Maybe T.Text
+      }
+
+instance ToJSON ScheduledEvent where
+  toJSON ScheduledEventExternal {..} =
+    Object
+      $ (if isJust scheduledEventExternalUserCount
+          then insert "user_count" (toJSON scheduledEventExternalUserCount)
+          else id
+        )
+      $ (if isJust scheduledEventExternalCreator
+          then insert "creator" (toJSON scheduledEventExternalCreator)
+          else id
+        )
+          base
+   where
+    Object base = object
+      [ "id" .= scheduledEventExternalId
+      , "guild_id" .= scheduledEventExternalGuildId
+      , "channel_id" .= Null
+      , "name" .= scheduledEventExternalName
+      , "creator_id" .= scheduledEventExternalCreatorId
+      , "description" .= scheduledEventExternalDescription
+      , "scheduled_start_time" .= scheduledEventExternalStartTime
+      , "scheduled_end_time" .= scheduledEventExternalEndTime
+      , "privacy_level" .= scheduledEventExternalPrivacyLevel
+      , "status" .= scheduledEventExternalStatus
+      , "entity_type" .= Number 3
+      , "entity_id" .= scheduledEventExternalEntityId
+      , "entity_metadata"
+        .= object ["location" .= scheduledEventExternalLocation]
+      , "creator" .= scheduledEventExternalCreator
+      , "user_count" .= scheduledEventExternalUserCount
+      , "image" .= scheduledEventExternalImage
+      ]
+  toJSON ScheduledEventStage {..} =
+    Object
+      $ (if isJust scheduledEventStageUserCount
+          then insert "user_count" (toJSON scheduledEventStageUserCount)
+          else id
+        )
+      $ (if isJust scheduledEventStageCreator
+          then insert "creator" (toJSON scheduledEventStageCreator)
+          else id
+        )
+          base
+   where
+    Object base = object
+      [ "id" .= scheduledEventStageId
+      , "guildId" .= scheduledEventStageGuildId
+      , "channel_id" .= scheduledEventStageChannelId
+      , "creator_id" .= scheduledEventStageCreatorId
+      , "name" .= scheduledEventStageName
+      , "description" .= scheduledEventStageDescription
+      , "scheduled_start_time" .= scheduledEventStageStartTime
+      , "scheduled_end_time" .= scheduledEventStageEndTime
+      , "privacy_level" .= scheduledEventStagePrivacyLevel
+      , "status" .= scheduledEventStageStatus
+      , "entity_type" .= Number 1
+      , "entity_id" .= scheduledEventStageEntityId
+      , "entity_metadata" .= Null
+      , "creator" .= scheduledEventStageCreator
+      , "user_count" .= scheduledEventStageUserCount
+      , "image" .= scheduledEventStageImage
+      ]
```

Comment:
The pattern elsewhere is to format ToJSON objects like the following. Allows mixing the `Maybe` values in with the always present values.

```haskell
instance ToJSON Attachment where
  toJSON Attachment {..} = object [(name, value) | (name, Just value) <-
      [ ("id",        toJSON <$> pure attachmentId)
      , ("filename",  toJSON <$> pure attachmentFilename)
      , ("size",      toJSON <$> pure attachmentSize)   -- always present so wrapped in pure == Just
      , ("url",       toJSON <$> pure attachmentUrl)
      , ("proxy_url", toJSON <$> pure attachmentProxy)
      , ("height",    toJSON <$>      attachmentHeight)
      , ("width",     toJSON <$>      attachmentWidth) -- Maybe, so fmapped with <$>
      ] ]
```


Helps with things like this PR's

```haskell
  $ (if isJust scheduledEventStageUserCount
      then insert "user_count" (toJSON scheduledEventStageUserCount)
      else id
    )
```

which I think would work like `, ("user_count",       toJSON <$> pure scheduledEventStageUserCount)`

Reply:
Oh I didnt see it then, that's fixed in the next commit